### DISCUSSION TITLE
How to apply two cut with in .arrays?

### DISCUSSION THREAD

**Discussion by ZhenxuanZhang-Jensen** [scikit-hep/uproot5#510] [Q&A] [answered]
Hello everyone:
I am trying to do some cuts in the .arrays, but no any description can do that with more than two cut apply.
for example:
events_250 = uproot.open("M250.root:Events")
and then apply two cuts on it:
events_250.arrays(cut = ("GEN_deltaR_MAX4Q>1.2" and "GEN_deltaR_MIN4Q<0.8"),library ="np")["GEN_deltaR_MIN4Q"]

**Comment by jpivarski** [comment, marked as answer]
You can write `cut = "(GEN_deltaR_MAX4Q>1.2) & (GEN_deltaR_MIN4Q<0.8)"` (with parentheses!), but also, there's no strong reason to put the cut in this string. You can cut the arrays after the `.arrays` call just as easily:

```python
ttree = uproot.open("M250.root:Events")
one_branch = ttree["GEN_deltaR_MIN4Q"]
GEN_deltaR_MAX4Q = one_branch.array(library ="np")
GEN_deltaR_MAX4Q_cut = GEN_deltaR_MAX4Q[(GEN_deltaR_MAX4Q>1.2) & (GEN_deltaR_MIN4Q<0.8)]
```

The `cut` argument does nothing more than put the string you give it into square brackets, which is how you cut in NumPy.

Also, calling `.arrays` without specifying any `expressions` or `filter_branches` reads all branches from disk. Putting `["GEN_deltaR_MIN4Q"]` after it picks out one of the already-read branches and throws away the rest. `.array` (above) is the easiest way to read only one branch, but you can also use `.arrays` with `expressions` or `filter_branches` to read a selection of branches.

It might not be relevant on small files, but on big files, reading unnecessary branches can be slow or impossible of you don't have enough memory.

**Reply by ZhenxuanZhang-Jensen** [reply]
Thank you for your answer. I still have confusion about how you apply the cut to all_events. it seems all_events is not even a variable or anything. The other thing is about the `.array` , you use one_branch.array only read one branch without any expressions?

**Reply by jpivarski** [reply]
Sorry, I fixed it above. I was trying to write it on my phone this morning, and I changed the code a few times for clarity.

In the above version, `one_branch` is a branch object and `GEN_deltaR_MAX4Q` is a NumPy array. If you pull a TBranch out of a TTree (`one_branch`), you can call [TBranch.array](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.TBranch.html#array) on it to get one array. If you don't, you can call [TTree.arrays](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TTree.TTree.html#arrays) on the TTree to get multiple arrays. With `library="np"`, it's a Python dict of NumPy arrays; with `library="ak"` (default), it's a single Awkward Array (with record structure instead of a dict, which is a bit more flexible); with `library="pd"`, it's a single Pandas DataFrame. See the links above for details.

But as for your original question, to get the "logical and" of multiple cuts, use the `&` operator with parentheses. (For the importance of parentheses, see #508.) It does not matter whether you do that inside the `cut` string or in Python after the array has been loaded. (A lot of people have been thinking that there's something special about putting the cut in the `cut` string and have been going to great lengths to express difficult things in it, when they could more easily be done in multiple lines of Python code. I want everyone to know that the only advantage `cut` provides is the convenience of putting the code inside square brackets for you!)
